src/main/java/picard/illumina/parser/TileMetricsUtil.java

```diff
@@ -117,18 +122,70 @@ private static File fileTileMetricsFile(File illuminaRunDirectory, Integer numCy
 
             //only create for cluster records
             if (record.isClusterRecord()) {
-                // Snag the phasing data for each read in the read structure. For both types of phasing values, this is the median of all of the individual values seen
-                final Collection<TilePhasingValue> tilePhasingValues = getTilePhasingValues(record.getLaneTileCode(),
-                        phasingMetricsFiles, readStructure, validationStringency);
-
-                tiles.add(new Tile(record.getLaneNumber(), record.getTileNumber(), tileMetricsIterator.getDensity(), record.getMetricValue(),
+                final Collection<TilePhasingValue> tilePhasingValues = phasingValues.get(record.getLaneNumber()).get(record.getTileNumber());
+                tiles.add(new Tile(record.getLaneNumber(), record.getTileNumber(), density, record.getMetricValue(),
                         tilePhasingValues.toArray(new TilePhasingValue[tilePhasingValues.size()])));
             }
         }
-
         return Collections.unmodifiableCollection(tiles);
     }
 
+    private static boolean v3TileMetricsHasClusterRecordsFirstPass(Map<String, ? extends Collection<IlluminaTileMetrics>> locationToMetricsMap) {
+        final Map.Entry<String, ? extends Collection<IlluminaTileMetrics>> entry = locationToMetricsMap.entrySet().iterator().next();
+        final IlluminaTileMetrics record = CollectionUtil.getSoleElement(entry.getValue());
+        return record.isClusterRecord();
+    }
+
+    public static Collection<Tile> parseClusterRecordsFromTileMetricsV3(
+            final Collection<File> tileMetricsOutFiles,
+            final Map<Integer, File> phasingMetricsFiles,
+            final ReadStructure readStructure
+    ) throws FileNotFoundException {
+        // Snag and cache the phasing data for each read in the read structure for each tile.
+        // For both types of phasing values, this is the median of all of the individual values seen
+        Map<Integer, Map<Integer, Collection<TilePhasingValue>>> phasingValues = getTilePhasingValues(phasingMetricsFiles, readStructure);
+        for (File tileMetricsOutFile : tileMetricsOutFiles) {
```

Comment:
For NovaSeq runs, multiple TileMetricsOut.bin files are generated. Theoretically we should be able to rely on the directory structure to decided which one to use, but this turned out to be unreliable. Ultimately, we want the file that contains cluster records ([file spec](http://illumina.github.io/interop/tile_v3.html)). In the worst case, making sure a file does not contain cluster records means parsing the file and iterating through each record it contains. This loop takes a quick peek into the file to try and find a cluster record (I believe a file may contain either just 'r' type records, just 't' type cluster records, or an equal number of both.). If it finds a cluster record, it collects and returns all cluster records in the current file. The next loop goes through every tile record in every tileMetricsFile until it has collected a non-zero number of records from a file, at which point it returns those tiles.

Since file parsing has now been significantly sped up, it may be the case that we aren't gaining much from this first pass. What do you think, @jacarey ?